src/cargo/util/toml/manifest_cache.rs

```diff
@@ -0,0 +1,102 @@
+use std::collections::{
+    hash_map::{Entry, HashMap},
+    BTreeSet,
+};
+use std::path::{Path, PathBuf};
+use std::rc::Rc;
+
+use super::{parse, TomlManifest};
+use crate::util::errors::{CargoResult, CargoResultExt, ManifestError};
+use crate::util::toml::{TomlProject, TomlWorkspace};
+use crate::util::{paths, Config};
+
+pub type ManifestCache = HashMap<PathBuf, ParseOutput>;
+
+#[derive(Debug, Clone)]
+pub struct ParseOutput {
+    pub manifest: Rc<TomlManifest>,
+    pub unused: Rc<BTreeSet<String>>,
+}
+
+impl ParseOutput {
+    pub fn workspace(&self) -> Option<&TomlWorkspace> {
+        self.manifest.workspace.as_ref()
+    }
+
+    pub fn package(&self) -> Option<&TomlProject> {
+        self.manifest
+            .package
+            .as_deref()
+            .or(self.manifest.project.as_deref())
+    }
+}
+
+pub fn parse_manifest<'a>(
```

Comment:
Could you add some comments to what this is doing? I didn't immediately understand what you meant by there was a circular dependency in manifest parsing in the PR description. I originally thought this was a performance-related cache but it sounds like this is required for correctness?

Reply:
I'll write a comment describing this but here is an explanation for now.

`do_read_manifest()` used to be self-contained:

    do_read_manifest()

Now, `do_read_manifest()` potentially needs the workspace root to define `MaybeWorkspace` values:

    do_read_manifest() -> find_root()

`find_root()` in turn tries to read the root workspace. A manifest can be its own root, so it forms a cycle:

    do_read_manifest() -> find_root() -> do_read_manifest() (-> find_root() -> do_read_manifest() -> ...)

In order to break this cycle, I added the `ManifestCache` which lets you shallow parse (`parse_manifest`) - i.e. deserializes & gives you access to the toml - without calling all of `do_read_manifest()`.

    do_read_manifest() -> find_workspace_root() -> parse_manifest()

This breaks the cycle while also ensuring we never parse the same manifest twice.